storage/shard.go

```diff
@@ -506,96 +547,107 @@ func (s *dbShard) tickAndExpire(
-				// This method does not run using a lock so this is safe to
-				// perform inline.
-				s.purgeExpiredSeries(expired)
-				for i := range expired {
-					expired[i] = nil
+			if err == series.ErrSeriesAllDatapointsExpired {
+				expired = append(expired, entry)
+				r.expiredSeries++
+				if len(expired) >= expireBatchLength {
+					// Purge when reaching max batch size to avoid large array growth
+					// and ensure smooth rate of elements being returned to pools.
+					// This method does not run using a lock so this is safe to
+					// perform inline.
+					s.purgeExpiredSeries(expired)
+					for i := range expired {
+						expired[i] = nil
+					}
+					expired = expired[:0]
+				}
+			} else {
+				r.activeSeries++
+				if err != nil {
+					r.errors++
 				}
-				expired = expired[:0]
 			}
-		} else {
-			r.activeSeries++
-			if err != nil {
-				r.errors++
+			r.activeBlocks += result.ActiveBlocks
+			r.openBlocks += result.OpenBlocks
+			r.wiredBlocks += result.WiredBlocks
+			r.unwiredBlocks += result.UnwiredBlocks
+			r.madeExpiredBlocks += result.MadeExpiredBlocks
+			r.madeUnwiredBlocks += result.MadeUnwiredBlocks
+			r.mergedOutOfOrderBlocks += result.MergedOutOfOrderBlocks
+			i++
+		}
+
+		if len(expired) > 0 {
+			// Purge any series that still haven't been purged yet
+			s.purgeExpiredSeries(expired)
+			for i := range expired {
+				expired[i] = nil
 			}
+			expired = expired[:0]
 		}
-		r.activeBlocks += result.ActiveBlocks
-		r.openBlocks += result.OpenBlocks
-		r.wiredBlocks += result.WiredBlocks
-		r.unwiredBlocks += result.UnwiredBlocks
-		r.madeExpiredBlocks += result.MadeExpiredBlocks
-		r.madeUnwiredBlocks += result.MadeUnwiredBlocks
-		r.mergedOutOfOrderBlocks += result.MergedOutOfOrderBlocks
-		i++
+
 		// Continue
 		return true
 	})
 
-	if len(expired) > 0 {
-		// Purge any series that still haven't been purged yet
-		s.purgeExpiredSeries(expired)
-	}
-
 	if terminatedTickingDueToClosing {
 		return tickResult{}, errShardClosingTickTerminated
 	}
 
 	return r, nil
 }
 
-func (s *dbShard) purgeExpiredSeries(expired []series.DatabaseSeries) {
+// NB(prateek): the calling function must ensure it has incremented the reference
+// count on the provided `expiredEntries`.
+func (s *dbShard) purgeExpiredSeries(expiredEntries []*dbShardEntry) {
 	// Remove all expired series from lookup and list.
 	s.Lock()
-	for _, series := range expired {
+	for _, entry := range expiredEntries {
+		series := entry.series
 		hash := series.ID().Hash()
 		elem, exists := s.lookup[hash]
 		if !exists {
 			continue
 		}
-		entry := elem.Value.(*dbShardEntry)
 		// If this series is currently being written to or read from, we don't
 		// remove it even though it's empty in that it might become non-empty
 		// soon.
 		// We avoid closing the series during a read to ensure a consistent view
 		// of the series if they manage to take a reference to it.
-		if entry.readerWriterCount() > 0 {
+		if entry.readerWriterCount() > 1 {
```

Comment:
I think @r asked about this too, but why 1 instead of 0?

Reply:
This addresses a bug in the original code: When iterating over a batch of elements, `forEachShardEntry' (and now `forEachShardEntryBatch`) inc the `readerWriterCount` of the element, and then dec the count once they're done with a batch. The original `tickAndExpire` called this function `purgeExpired...` within the function passed to `forEachShardEntry`. Any elements in the `expired` list being purged, which were in the current underlying batch used by `forEachShardEntry` were guaranteed to have a readerWriterCount > 0 (at least 1) and we'd skip expiring them pessimistically. TBH: the original code maintained the `expired` list across `forEachShardEntry` calls, so it wasn't entirely fked. 

The new code maintains `expired` only within each underlying batch used by `forEachShardEntryBatch`. This is why we can use the `> 1` check and be more accurate about cleanup.

Reply:
I'm still not totally following, but could you add a comment explaining it? It looks like something that would be tempting for someone to "fix" right now

Reply:
sure thing

Reply:
Put down a verbose comment and a failing if block to explicitly callout the non-expected condition. Does it make more sense now?